src/main/java/org/prebid/server/bidder/criteo/CriteoBidder.java

```diff
@@ -0,0 +1,293 @@
+        final ExtRegs extRegs = getExtRegs(bidRequest);
+        criteoRequestBuilder
+                .publisher(buildCriteoPublisher(bidRequest, getNetworkId(bidRequest)))
+                .user(buildCriteoUser(bidRequest, extRegs))
+                .gdprConsent(buildCriteoGdprConsent(bidRequest, extRegs));
+
+        if (bidRequest.getUser() != null && bidRequest.getUser().getExt() != null) {
+            final ExtUser extUser = bidRequest.getUser().getExt();
+            criteoRequestBuilder.eids(extUser.getEids());
+        }
+
+        final CriteoRequest criteoRequest = criteoRequestBuilder.build();
+        final String requestBody;
+        try {
+            requestBody = jsonMapper.encode(criteoRequest);
+        } catch (EncodeException e) {
+            return Result.withError(BidderError.badInput(
+                    String.format("Failed to encode request body, error: %s", e.getMessage())));
+        }
+
+        return Result.withValue(HttpRequest.<CriteoRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .body(requestBody)
+                .headers(resolveHeaders(criteoRequest))
+                .payload(criteoRequest)
+                .build());
+    }
+
+    private CriteoRequestSlot buildRequestSlot(Imp imp) {
+        final CriteoRequestSlot.CriteoRequestSlotBuilder criteoRequestSlotBuilder
+                = CriteoRequestSlot.builder()
+                .impId(imp.getId())
+                .slotId(criteoSlotIdGenerator.generateUuid())
+                .sizes(getImpSizesFromBanner(imp.getBanner()));
+
+        final ExtImpCriteo extImpCriteo = parseImpExt(imp);
+        if (extImpCriteo.getZoneId() != null && extImpCriteo.getZoneId() > 0) {
+            criteoRequestSlotBuilder.zoneId(extImpCriteo.getZoneId());
+        }
+
+        if (extImpCriteo.getNetworkId() != null && extImpCriteo.getNetworkId() > 0) {
+            criteoRequestSlotBuilder.networkId(extImpCriteo.getNetworkId());
+        }
+        return criteoRequestSlotBuilder.build();
+    }
+
+    private static List<String> getImpSizesFromBanner(Banner banner) {
+        if (banner == null) {
+            return new ArrayList<>();
+        }
+
+        final List<String> sizes = new ArrayList<>();
+        if (banner.getFormat() != null) {
+            for (Format format : banner.getFormat()) {
+                sizes.add(String.format("%sx%s", format.getW(), format.getH()));
+            }
+        } else if (banner.getW() != null && banner.getW() > 0
+                && banner.getH() != null && banner.getH() > 0) {
+            sizes.add(String.format("%sx%s", banner.getW(), banner.getH()));
+        }
+        return sizes;
+    }
+
+    private ExtImpCriteo parseImpExt(Imp imp) {
+        try {
+            return jsonMapper.mapper().convertValue(imp.getExt().get("bidder"), ExtImpCriteo.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
```

Comment:
This exception should be covered by caller's code.
The idea is that `makeHttpRequests` method never throws exceptions.

Reply:
catch() clause at caller added